tubular/dates.py

```diff
@@ -1040,20 +1040,56 @@ def __init__(
                 )
             )
 
-        if not isinstance(units, str):
+        if not isinstance(units, str) and not isinstance(units, dict):
             raise TypeError(
-                "{}: units must be a string but got {}".format(
+                "{}: units must be a string or dict but got {}".format(
                     self.classname(), type(units)
                 )
             )
 
-        if (not isinstance(period, int)) and (not isinstance(period, float)):
+        if (
+            (not isinstance(period, int))
+            and (not isinstance(period, float))
+            and (not isinstance(period, dict))
+            or (isinstance(period, bool))
+        ):
             raise TypeError(
-                "{}: period must be a int or float but got {}".format(
+                "{}: period must be an int, float or dict but got {}".format(
                     self.classname(), type(period)
                 )
             )
 
+        if isinstance(units, dict):
+            if not all(isinstance(item, str) for item in list(units.keys())) or not all(
+                isinstance(item, str) for item in list(units.values())
+            ):
+                raise TypeError(
+                    "{}: units dictionary key value pair must be strings but got {} {}".format(
+                        self.classname(),
+                        set(type(k) for k in units.keys()),
+                        set(type(v) for v in units.values()),
+                    )
+                )
+
+        if isinstance(period, dict):
+            if (
+                not all(isinstance(item, str) for item in list(period.keys()))
+                or (
+                    not all(isinstance(item, int) for item in list(period.values()))
+                    and not all(
+                        isinstance(item, float) for item in list(period.values())
+                    )
+                )
+                or any(isinstance(item, bool) for item in list(period.values()))
+            ):
+                raise TypeError(
+                    "{}: period dictionary key value pair must be str:int or str:float but got {} {}".format(
+                        self.classname(),
+                        set(type(k) for k in period.keys()),
```

Comment:
error message might be hard to read as dtype for all keys presented then dtype of all values rather than in pairs as they will ahve been input.  Either specify keys: {} values: {} in the error message or find a way to present the dtypes in a format more similar to input.